iib/web/models.py

```diff
@@ -190,6 +192,93 @@ def add_state(self, state, state_reason):
         db.session.flush()
         self.request_state_id = request_state.id
 
+    def to_json(self, verbose=True):
+        def _state_to_json(state):
+            return {
+                'state': RequestStateMapping(state.state).name,
+                'state_reason': state.state_reason,
+                'updated': state.updated.isoformat() + 'Z',
+            }
+
+        states = [_state_to_json(state) for state in self.states]
+        # Reverse the list since the latest states should be first
+        states = list(reversed(states))
+        latest_state = states[0]
+
+        rv = {
+            'id': self.id,
+            'arches': [arch.name for arch in self.index_image.architectures],
+            'binary_image': self.binary_image.pull_specification,
+            'binary_image_resolved': self.binary_image_resolved.pull_specification,
+            'bundles': [bundle.pull_specification for bundle in self.bundles],
+            'from_index': self.from_index.pull_specification,
+            'from_index_resolved': self.from_index_resolved,
+            'index_image': self.index_image.pull_specification,
+            'state': latest_state,
+            'user': self.user.username,
+        }
+
+        if verbose:
+            rv['state_history'] = states
+
+        return rv
+
+    @classmethod
+    def from_json(cls, kwargs):
+        # Validate all required parameters are present
+        required_params = {'bundles', 'binary_image'}
+        optional_params = {'from_index', 'add_arches', 'cnr_token'}
+        missing_params = required_params - set(kwargs.keys()) - optional_params
```

Comment:
Optional: In Python 3, `.keys()` produces an object that can be used in set operations, so you don't need to cast it to a set.